docs/rfcs/20220131_raft_snapshots.md

```diff
@@ -0,0 +1,149 @@
+
+To fix existing problems which prevents transacitons from working with the compacted topics. To introduce infastructure to support custom (probably WASM based) state machines.
+
+## What use cases does it support?
+
+ - Idempotency
+ - Transactions
+ - Offset translator
+ - Custom state machines
+
+# Guide-level explanation
+## Context
+
+Some of the Redpanda operations require knowledge about all records passed through the partition (Raft log) up to this moment (let's call them aggregate-dependent).
+ 
+ - Idempotency: Redpanda inspects the messages to update the list of ongoing sessions and then uses it to reject the out of order requests.
+ - Offset translator: Redpanda tracks the reconfiguration batches to adjust raft offsets into Kafka offsets.
+ - Transactions: Redpanda inspects the records to keep an eye on the ongoing transactions and maintain the right consistency level
+ 
+At the same time disks have limited capacity and data eventually lose actuality so Redpanda supports several cleanup policies:
+ 
+ - Deletion
+ - Compaction
+ 
+Deletion clean up policy removes some messages (defined by age or size) from the head of the log.
+
+![](./imgs/20220131_log.png)
+
+Compact clean up policy is a bit more complicated: each record in Redpanda has a key and with compaction enabled Redpanda keeps a record at the tail of the log but removes some or all records with the same key written before it.
+
+## Problem
+
+There is a problem with the integration of aggregate-dependent operations and the cleanup policies: Redpanda may remove data before the operation sees it (it's especially true during reconfiguration because the new node never saw a non compacted log).
+
+![](./imgs/20220131_cleanup.png)
+
+Imagine that Redpanda stones withdraw & deposit history for a set of accounts and an aggregate-dependent operation needs an account balance.
+
+![](./imgs/20220131_compaction.png)
+
+We see that the state after compaction is completely different: `(2,-2,4)` vs `(3,1,4)`. The only way to guarantee the correct result is to cache state before the compaction / deletion.
+
+Local caching isn't enough. Raft tolerates note failures as long as majority is connected meaning a single node may fall behind the majority and behind the compaction horizon.
+
+![](./imgs/20220131_behind.png)
+
+When the isolated node rejoins the cluster it fetches an already compacted log with holes and the state diverges.
+
+# Reference-level explanation
+
+## Approach #1: Ship the snapshot (cached state) via the log
+
+The main idea is to use existing log replication to ship the snapshot. It's easier to implement but just like the reconfiguration batches it shifts the offsets and makes us continue to rely on offset translation.
+
+### Overview
+
+A leader decides to do compaction at offset X:
+ 
+  1. saves current snapshot
+  2. waits until it's safe to do compaction:
+    a. segment containing X is closed
+    b. tx front is passed the end of the segment
```

Comment:
I'm not sure what this means

Reply:
`a` and `b` are conditions which should be met to perform compaction safely.

Segment is a collection of records, each record has an offset, before Redpanda performs a compaction it should wait until a record collection containing a record with offset X is filled then we close it and start another collection.

Each transactions has a start offset, tx front is a set of start offsets of all ongoing transacitons. Tx front is passed the end of the segment means that all start offsets of the ongoing transactions are greater the the last offset of the segment.

Reply:
You can change /passed/past/ here. Also we don't define tx front here, so maybe expand it:

> All ofsets in set of in-flight transactions, `tx front`, are past the end of the segment.